### PR TITLE
fermyon-spin 1.4.1 (new formula)

### PR DESCRIPTION
**Pull request by kate-goldenring**
Add a Formula for Spin, an open-source tool for building and running serverless WebAssembly applications.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by kate-goldenring** [comment]
I am a little stuck debugging the CI issues and would appreciate any help debugging. I am not able to reproduce them locally on MacOS 13 arm or Linux (Ubuntu 20.04). Currently a permissions error is being thrown during the `post_install` step:

```sh
    Errno::EACCES: Permission denied @ rb_sysopen - /opt/homebrew/etc/fermyon-spin/plugins/.spin-plugins/.git/objects/pack/pack-02bf44b8e94042809d82707294ef560e741ed85b.idx
```

I tried added more permissions to the `/opt/homebrew/etc/fermyon-spin/plugins/.spin-plugins` directory but it is still occurring. `/opt/homebrew/etc/fermyon-spin/plugins/.spin-plugins` is where a Git repository of available Spin plugins is setup during installation.

File: Formula/fermyon-spin.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+class FermyonSpin < Formula
+  desc "Open-source tool for building and running serverless WebAssembly applications"
+  homepage "https://spin.fermyon.dev/"
+  url "https://github.com/fermyon/spin.git",
+    tag:      "v1.3.0",
+    revision: "9fb8256d1380a046414b22bf2c17d1543f5029e4"
+  license "Apache-2.0" => { with: "LLVM-exception" }
+
+  depends_on "rust" => ["1.69.0", :build, :test]
```
Comment (chenrui333) [inline Formula/fermyon-spin.rb:9]: this should be good
```suggestion
  depends_on "rust" => [build, :test]
```

**Comment by kate-goldenring** [comment]
Currently failing on linux with:
```sh
`brew linkage --test fermyon-spin` failed on Linux!

Unwanted system libraries:
  /lib/x86_64-linux-gnu/libz.so.1
```

I added `uses_from_macos "zlib"` as recommended in other PRs but it is still failing. It may be due to the default features we are using in our `flake` dependency

File: Formula/fermyon-spin.rb [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    system "#{bin}/spin", "plugins", "install", "js2wasm", "--yes"
+    system "#{bin}/spin", "plugins", "install", "py2wasm", "--yes"
+    system "#{bin}/spin", "plugins", "install", "cloud", "--yes"
+    # Set permissions for local plugins repository
+    chmod_R(0755, spin_datadir/"plugins")
+  end
+
+  def spin_datadir
+    pkgetc
+  end
```
Comment (SMillerDev) [inline Formula/fermyon-spin.rb:39]:
```suggestion
    chmod_R(0755, pkgetc/"plugins")
  end
```

File: Formula/fermyon-spin.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  depends_on "rustup-init" => [:build, :test]
+
+  uses_from_macos "zlib"
+
+  on_linux do
+    depends_on "pkg-config" => :build
+  end
+
+  def install
+    system "#{Formula["rustup-init"].bin}/rustup-init", "-qy", "--no-modify-path"
```
Comment (SMillerDev) [inline Formula/fermyon-spin.rb:19]: This should be build with `rust` instead

**Comment by kate-goldenring** [comment]
@SMillerDev do you have an idea of why Linux is detecting the unwanted `zlib` library despite it being added as a dependency?

File: Formula/fermyon-spin.rb (hunk 3 above)
Comment (SMillerDev) [reply Formula/fermyon-spin.rb:19]: Why wouldn't the normal rust do this?

Comment (kate-goldenring) [reply Formula/fermyon-spin.rb:19]: @SMillerDev `rust` cannot add targets and it doesn't seem that wasm targets are packaged with the homebrew `rust` formula. Using `rustup` is the next best bet and does appear to be a convention, given [`rustfmt`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/rustfmt.rb) and [`cargo-zigbuild`](https://github.com/Homebrew/homebrew-core/blob/master/Formula/cargo-zigbuild.rb#L36)

**Comment by kate-goldenring** [comment]
@carlocab @SMillerDev do you have any recommendations for handling the unwanted [`zlib` dependency](https://github.com/Homebrew/homebrew-core/pull/133861#issuecomment-1599923000)? Or, for now, should i make the formula MacOS only, and we can embed the zlib library in Spin next release?

**Comment by kate-goldenring** [comment]
@carlocab @SMillerDev I am running into a `brew audit` issue for using `rustup-init` is there a path forward for me here given that I need to add the `wasm32-wasi` target in order to build Spin?

```
brew audit fermyon-spin --online --new-formula
fermyon-spin
  * line 10, col 3: Formulae in homebrew/core should use 'depends_on "rust" => [:build, :test]' instead of 'depends_on "rustup-init" => [:build, :test]'.
  * line 20, col 12: Formula in homebrew/core should not use `rustup-init` at build-time.
```

Support would be appreciated. Please let me know if it will not be possible to get Spin into Homebrew core and i should create a private tap.

**Comment by SMillerDev** [comment]
I think using a private tap would be better, downloading a binary Rust release just does not fit the model for Homebrew/core very well.

**Comment by carlocab** [comment]
Alternatively, you can try adding WASM support to the existing Rust formula, since that seems to be what you need here.

**Pull request closed without merging**